### PR TITLE
#76 - Better break-up fraction. 

### PR DESCRIPTION
**Pull request by eltonlaw**
#76 This is only solves the initial case posed of (2x+3)/(2x+2). The output is now
(2x+2)/(2x+2) + (1)/(2x+2).

### CODE REVIEW

File: lib/simplifyExpression/breakUpNumeratorSearch/index.js [hunk 1]
```diff
@@ -27,8 +28,33 @@ function breakUpNumerator(node) {
   // At this point, we know that node is a fraction and its numerator is a sum
   // of terms that can't be collected or combined, so we should break it up.
   const fractionList = [];
-  const denominator = node.args[1];
-  numerator.args.forEach(arg => {
+  let denominator = node.args[1];
+
+  // Check if we can add a constant to make the fraction nicer
+  // fraction e.g. (2+x)/(5+x) -> (5+x)/(5+x) + 3/(5+x)
```
Comment (aliang8) [inline lib/simplifyExpression/breakUpNumeratorSearch/index.js]: I think you mean - 3/ (5 + x) *.

File: lib/checks/canFindDenominatorInNumerator.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    numerator = numerator.content;
+  }
+  if (Node.Type.isParenthesis(denominator)) {
+    denominator = denominator.content;
+  }
+
+  if (numerator.op !== '+') {
+    return false;
+  }
+  if (!('args' in denominator)) {
```
Comment (aliang8) [inline lib/checks/canFindDenominatorInNumerator.js]: I'm not understanding what this part does. Can you explain?

File: lib/checks/canFindDenominatorInNumerator.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  let numerator = node.args[0];
+  let denominator = node.args[1];
+  if (Node.Type.isParenthesis(numerator)) {
+    numerator = numerator.content;
+  }
+  if (Node.Type.isParenthesis(denominator)) {
+    denominator = denominator.content;
+  }
+
+  if (numerator.op !== '+') {
```
Comment (aliang8) [inline lib/checks/canFindDenominatorInNumerator.js]: Try adding the case where the numerator has minus operation too.

File: lib/checks/canFindDenominatorInNumerator.js (hunk 2 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: line 23 is to check whether the numerator is the '+' op between nodes, so that's to check that there's an 'x + y' kind of thing going on, but now that you mention it, it's not actually required. cause we should be able to go from x/(x+1) -> (x+1)/(x+1) -1)/(x+1). I'll take this out.

line 26 is in there because tests were failing by having a constant in the bottom, i think a constant in the bottom is resolved best the original way.

File: lib/simplifyExpression/breakUpNumeratorSearch/index.js (hunk 1 above)
Comment (eltonlaw) [inline lib/simplifyExpression/breakUpNumeratorSearch/index.js]: Whoops, yeah haha

File: lib/checks/canFindDenominatorInNumerator.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+const Node = require('../node');
+
+// Returns true if by adding a term you can simplify part of the function into an integer
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I know it's almost 80 characters here so it feels annoying to put it in two lines, but... 80 character max please ^_^

File: lib/checks/canFindDenominatorInNumerator.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+// e.g. (2x+1)/(2x+3) -> True
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: it would be easiest to understand this function (from an outsider perspective) if you explain why it's true as well

```
// e.g. (2x+1)/(2x+3) -> True 
// because of the simplification  (2x+1)/(2x+3) -> (2x + 3)/(2x + 3) - 2/(2x + 3) -> 1 - 2/(2x + 3) 
```

File: lib/checks/canFindDenominatorInNumerator.js (hunk 2 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: > actually I think with this we probably need to check that there's some sort of subtraction or addition somewhere in the fraction, so maybe set it to try to catch any of those scenarios. What do you think?

I think the best way to check this is to make sure that the numerator and denominator aren't equal. if they're both `x` then they're equal and there's nothing to do. but also if they're both `x+1` then this step doesn't apply, we can just cancel (that step's code is elsewhere). How's that sound?

File: lib/checks/canFindDenominatorInNumerator.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    denominator = denominator.content;
+  }
+
+  if (numerator.op !== '+') {
+    return false;
+  }
+  if (!('args' in denominator)) {
+    return false;
+  }
+  const numeratorFirstTerm = new Node.PolynomialTerm(numerator.args[0]);
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: most of the time I think it'll be sorted so that the first term is the poly term, but I don't think this is guaranteed. you could have `(1+x) / (x+2)`. So we might need to have a step somewhere that sorts, or otherwise iterate through the arguments to find this poly term

Also, where do you make sure they have the same exponent? (i.e. for now that it's always 1)

File: lib/simplifyExpression/breakUpNumeratorSearch/index.js (hunk 1 above)
Comment (evykassirer) [inline lib/simplifyExpression/breakUpNumeratorSearch/index.js]: so then you'd also want to say add/subtract ^_^

File: lib/simplifyExpression/breakUpNumeratorSearch/index.js [hunk 7]
```diff
@@ -27,8 +28,33 @@ function breakUpNumerator(node) {
...
+  if (canFindDenominatorInNumerator(node)) {
+    let denominatorParenRemoved = false;
+    if (Node.Type.isParenthesis(denominator)) {
+      denominatorParenRemoved = true;
+      denominator = denominator.content;
+    }
+    const newNumerator = [];
+
+    // The constant value difference between the numerator and the denominator
+    const numeratorConstant = parseInt(numerator.args[1].value);
```
Comment (evykassirer) [inline lib/simplifyExpression/breakUpNumeratorSearch/index.js]: I don't think you ever make sure that this is actually a constant - do a check either here or in `canFindDenominatorInNumerator` or else you'll get an error here

File: lib/simplifyExpression/breakUpNumeratorSearch/index.js [hunk 8]
```diff
@@ -27,8 +28,33 @@ function breakUpNumerator(node) {
...
+
+    newNumerator.push(Node.Creator.constant(addedConstant));
+    newNumerator.push(denominator);
+
+    numerator = newNumerator;
+
+    if (denominatorParenRemoved) {
+      denominator = Node.Creator.parenthesis(denominator);
+    }
+  }
```
Comment (evykassirer) [inline lib/simplifyExpression/breakUpNumeratorSearch/index.js]: I think it'd be best if we can break this up into two steps. the step you added, and the break up fraction that comes after (and then maybe we can group in the cancelling that comes after). and then we'd have substeps for breaking up the fraction.

what do you think? (let me know if you want me to explain more what this means)

File: test/checks/checks.test.js [hunk 9]
```diff
@@ -29,3 +29,10 @@ describe('canSimplifyPolynomialTerms addition', function() {
   ];
   tests.forEach(t => testCanCombine(t[0], t[1]));
 });
+
+describe('canSimplifyPolynomialTerms denominator in numerator', function() {
+  const tests = [
+    ['(2x + 3)/(2x + 2)', true],
```
Comment (evykassirer) [inline test/checks/checks.test.js]: can you add a bunch more tests? especially a cases that should be false

you can use some of my comments above for edge cases to test for too :)

File: test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js [hunk 10]
```diff
@@ -11,6 +11,7 @@ describe('breakUpNumerator', function() {
     ['(x+3+y)/3', '(x / 3 + 3/3 + y / 3)'],
     ['(2+x)/4', '(2/4 + x / 4)'],
     ['2(x+3)/3', '2 * (x / 3 + 3/3)'],
+    ['(2x + 3)/(2x + 2)', '(1 / (2x + 2) + (2x + 2) / (2x + 2))'],
```
Comment (evykassirer) [inline test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js]: same here - try to think of edge cases. I can help you brainstorm if you dunno what to add

**Comment by aliang8** [comment]
Hi @eltonlaw and @evykassirer,
I found a really cool approach to long polynomial division that I would like to share with the two of you. It's super intuitive, pedagogical, just like how we would normally do long division. 

The pseudocode is here: https://rosettacode.org/wiki/Polynomial_long_division#Python

Basically, it represents the polynomials as vectors such that the i^th element keeps the coefficient of x^i
e.g [-42 0 -12 1] => x^3 - 12 x^2 + 0x - 42
and multiplication by a monomial is a shift of the vector's element to the right
e.g [-42 0 -12 1] * x^2 => [0 0 -42 0 -12 1]
It uses a simple recursive loop to find all the coefficients of the quotient polynomial by applying the two concepts above.
Highly suggest taking a look at it since it covers most if not all of the edge cases. Let me know what yall think of it 👍

File: lib/checks/canFindDenominatorInNumerator.js [hunk 11]
```diff
@@ -19,17 +20,42 @@ function canFindDenominatorInNumerator(node) {
...
+        denominator.op === '+' || numerator.op === '-')) {
+    return false;
+  }
+  if (denominator.op !== '+') {
+    return false;
+  }
+
+  let numeratorFirstTerm;
+  if (numerator.op === '+') {
+    numeratorFirstTerm = new Node.PolynomialTerm(numerator.args[0]);
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I'm concerned that we can just assume it'll be a polynomial term

(what if the second arg is a polynomial term? what if there are more than 2 args?)

Sorry this got more complicated haha. If you want, we can chat on gitter or something and figure out how to attack this

File: lib/checks/canFindDenominatorInNumerator.js [hunk 12]
```diff
@@ -19,17 +20,42 @@ function canFindDenominatorInNumerator(node) {
...
-  if (numerator.op !== '+') {
+  if (!(numeratorFirstTerm)) {
     return false;
   }
-  if (!('args' in denominator)) {
+  if (!(denominatorFirstTerm)) {
+    return false;
+  }
+
+  if (!(numeratorFirstTerm.getSymbolName() === 'x' && denominatorFirstTerm.getSymbolName() === 'x')) {
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I think it's better to just say they have the same symbol. Then if they share the symbol `y`, it'd work too!

File: lib/checks/canFindDenominatorInNumerator.js [hunk 13]
```diff
@@ -19,17 +20,42 @@ function canFindDenominatorInNumerator(node) {
   if (Node.Type.isParenthesis(denominator)) {
     denominator = denominator.content;
   }
+  if (!(numerator.op === '+' || numerator.op === '-' ||
+        denominator.op === '+' || numerator.op === '-')) {
```
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: whoops that should be denominator.op

File: lib/checks/canFindDenominatorInNumerator.js [hunk 14]
```diff
@@ -19,17 +20,42 @@ function canFindDenominatorInNumerator(node) {
...
+    return false;
+  }
+  if (denominator.op !== '+') {
```
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: I was thinking we don't need to use this extended version on it. The original breakUpNumerator function works fine for this case. But I think you've highlighted an interesting point, anyone reading the code would be confused.

File: lib/checks/canFindDenominatorInNumerator.js [hunk 15]
```diff
@@ -19,17 +20,42 @@ function canFindDenominatorInNumerator(node) {
   if (Node.Type.isParenthesis(denominator)) {
     denominator = denominator.content;
   }
+  if (!(numerator.op === '+' || numerator.op === '-' ||
```
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: Okay

File: lib/checks/canFindDenominatorInNumerator.js (hunk 11 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: Yeah, this is what I wanted to talk about actually. One of the main issues is that it's not sorted, so I wanted to ask you all if we should add a sorting function which will run before this. Or if you don't see any merit in that, we could add some logic to find the highest polynomial term right into the function (or outside of it if you want).

File: lib/checks/canFindDenominatorInNumerator.js (hunk 12 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: oh true, I didn't think about the symbol being a different name. The reason I did it to equal 'x' was to restrict it to the case of x^1

File: test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js [hunk 16]
```diff
@@ -11,7 +11,13 @@ describe('breakUpNumerator', function() {
     ['(x+3+y)/3', '(x / 3 + 3/3 + y / 3)'],
     ['(2+x)/4', '(2/4 + x / 4)'],
     ['2(x+3)/3', '2 * (x / 3 + 3/3)'],
-    ['(2x + 3)/(2x + 2)', '(1 / (2x + 2) + (2x + 2) / (2x + 2))'],
+    ['(2x + 3)/(2x + 2)', '((2x + 2) / (2x + 2) + 1 / (2x + 2))'],
+    ['(2x - 3)/(2x + 2)', '((2x + 2) / (2x + 2) - 5 / (2x + 2))'],
+    ['(2x + 3)/(2x)', '(2x / (2x) + 3 / (2x))'],
+    ['(3 + 2x)/(2x)', '(3 / (2x) + 2x / (2x))'],
+    ['(4x + 3)/(2x + 2)', '(2 * (2x + 2) / (2x + 2) - 1 / (2x + 2))'],
+//    ['(2x)/(3 + 2x)', '((3 + 2x) / (3 + 2x) - 3 / (3 + 2x))'],
```
Comment (eltonlaw) [inline test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js]: As mentioned earlier, I didn't know what to do without some sort of way to sort it.

File: lib/checks/canFindDenominatorInNumerator.js (hunk 11 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: yeah sure, I'm down to chat on gitter

File: test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js (hunk 16 above)
Comment (evykassirer) [inline test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js]: ah, cool - sorry for missing that

File: lib/checks/canFindDenominatorInNumerator.js (hunk 12 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: yeah I think just doing getSymbolName()  === getSymbolName()  would work!

File: lib/checks/canFindDenominatorInNumerator.js (hunk 14 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: ooo true. Maybe just add a comment saying where the other case is handled then!

File: lib/checks/canFindDenominatorInNumerator.js (hunk 11 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: sweet - okay I actually am busy all day tomorrow >.> (first day of work!!) 

does Wed evening work for you?

Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: or can we chat asynchronously - I can try to think about solutions to this tomorrow

File: lib/checks/canFindDenominatorInNumerator.js (hunk 12 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: wait I'm not really getting the use of getSymbolName() actually, what other values can the symbol even take, it seems like it always returns 'x'. Is this just for when there are multiple variables? Cause that seems way out of the scope for now anyways. Would it be better to check that denominatorFirstTerm.getExponentNode() === numeratorFirstTerm.getExponentNode() or check if they equal undefined?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 11 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: uhh yeah lets just chat asynchronously then, no worries take your time, good luck!

File: lib/checks/canFindDenominatorInNumerator.js (hunk 12 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: using `getSymbolName` makes sure the symbol in numerator and denominator are the same - i.e. both `x` or both `y` or whatever - you could take any example with `x`s and change them to `y`s and it should also work (actually can you add a test for that? ^_^)

`numeratorFirstTerm.getSymbolName() === denominatorFirstTerm.getSymbolName()` is what I was thinking

File: lib/checks/canFindDenominatorInNumerator.js (hunk 11 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I think you could start with doing this rule _only_ if there are exactly one or two args, the first is always a polynomial term, and the second if it exists is always a constant

i.e. come up with very limited cases and then test for them before going forward

and then after merging this you could add some more cases if they're easy enough

what do you think?

Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: oh whoops didn't see this, sure sounds good!

**Comment by eltonlaw** [comment]
I think I incorporated all your feedback @evykassirer, as things are now is this okay to get merged?

File: lib/checks/canFindDenominatorInNumerator.js [hunk 17]
```diff
@@ -20,38 +20,69 @@ function canFindDenominatorInNumerator(node) {
...
     return false;
   }
-  if (denominator.op !== '+') {
+  if (!(denominator.op === '+' || denominator.op === '-')) {
     return false;
   }
-
+  // Check if numerator's second argument is a constant if numerator has two arguments
+  if (n_args_length === 2) {
+    if (!(Node.Type.isConstant(numerator.args[1]))) {
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: you don't need the parens around Node.Type(.....)

File: lib/checks/canFindDenominatorInNumerator.js [hunk 18]
```diff
@@ -20,38 +20,69 @@ function canFindDenominatorInNumerator(node) {
...
-  }
-  if (!(denominatorFirstTerm)) {
+  // If an exponent exists (aka not x^1), return false
+  if (numeratorFirstTerm.getExponentNode() ||
+      denominatorFirstTerm.getExponentNode()) {
     return false;
   }
-
-  if (!(numeratorFirstTerm.getSymbolName() === 'x' && denominatorFirstTerm.getSymbolName() === 'x')) {
+  // Check that the symbols are the same, Ex. (x+1)/(y+1) would not pass
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I love the example in the comment :D

File: lib/checks/canFindDenominatorInNumerator.js [hunk 19]
```diff
@@ -20,38 +20,69 @@ function canFindDenominatorInNumerator(node) {
...
+  if ('args' in denominator) {
+    d_args_length = denominator.args.length;
+  }
+  else {
+    d_args_length = 1;
+  }
+
+  // If numerator isn't length 2 or length 1 with a polynomial return false
+  if (!(n_args_length === 2)) {
+    if (!(n_args_length === 1 || Node.Type.isConstant(numerator))) {
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: also you should be checking for polynomial, not constant, right?

Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: did you also check that the first argument of the numerator/denominator is a polynomial term?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 17 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: oh true!

File: lib/checks/canFindDenominatorInNumerator.js [hunk 20]
```diff
@@ -20,38 +20,69 @@ function canFindDenominatorInNumerator(node) {
   if (Node.Type.isParenthesis(denominator)) {
     denominator = denominator.content;
   }
-  if (!(numerator.op === '+' || numerator.op === '-' ||
-        denominator.op === '+' || numerator.op === '-')) {
+
+  let n_args_length;
+  // If numerator is '*' op, it signifies a single 'ax', should be assigned a
+  // length of 1
+  if ('args' in numerator && numerator.op !== '*') {
```
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: ah yeah good catch, it should be ===

File: lib/checks/canFindDenominatorInNumerator.js [hunk 21]
```diff
@@ -20,38 +20,69 @@ function canFindDenominatorInNumerator(node) {
...
+  }
+
+  // If numerator isn't length 2 or length 1 with a polynomial return false
+  if (!(n_args_length === 2)) {
+    if (!(n_args_length === 1 || Node.Type.isConstant(numerator))) {
+      return false;
+    }
+  }
+  // Function doesn't support denominators with args > 2
+  // Tf d_args_length < 2 the normal functionality already covers it
```
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: normal functionality is just taking each arg of the numerator and putting a denominator under it. So (ax^2 +bx +c)/d decomposes to ax^2/d +bx/d + c/d

Yeah it's the same as 1, I'll change it

File: lib/checks/canFindDenominatorInNumerator.js (hunk 19 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: You're right, it should be a check for polynomial, originally I wanted to put isSymbol() there, but it doesn't work on values that have a coefficient: ex. Node.Type.isSymbol('2x') evaluates to false, so i figured the logical positive would work instead: !Node.Type.isConstant('2x') evaluates to true.

Also, I don't check that the first argument of the numerator/denominator is a polynomial term because I assume that if we've already evaluated through and there's a second arg in the numerator/denominator that's a constant, the only thing left in the first args MUST be a polynomial, cause the other simplifiers would have added like terms already. Do you want me to explicitly write !Node.Type.isConstant(denominator.arg[0])?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 21 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: can you give that example in the comment just so people know what you mean?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 20 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: great - make sure the comment explains why (not just what - so not just saying what the if statement is looking for, but an example like what you showed me so people know why you're checking that)

File: lib/checks/canFindDenominatorInNumerator.js (hunk 19 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: there's a `isPolynomialTerm` (or something similar) defined somewhere - I think that's what you want. You could actually probably use it above when your'e trying to avoid the ax case (just make sure it's not a poly term)

`!Node.Type.isConstant('2x')` isn't rigorous enough because there are a bunch (5?) different types of node types - you can look in `Node/Type` to see them all

a counterexample to your assumption - what if it was `5/(2^x + 3)`?

**Comment by evykassirer** [comment]
I'm at GoogleIO this week and can't take much time to review your code - but I replied to some comments and will actually look at the diff on the weekend

thanks for your patience! soon these rules will be easier to write hopefully - someone's working on a cleaner way to describe them right now :) (if you want to see more, check out https://github.com/semantic-math/math-rules)

File: lib/checks/canFindDenominatorInNumerator.js [hunk 22]
```diff
@@ -5,8 +5,21 @@ const Node = require('../node');
 // e.g. (2x+1)/(2x+3) -> True because of the following simplification
 // (2x+1)/(2x+3) -> (2x + 3)/(2x + 3) - 2/(2x + 3) -> 1 - 2/(2x + 3)
 // e.g. (2x+1)/(2x^2 + 3) -> False
+// =========================================================================
+// CHECKS
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I think it would be most clear if you explain what exactly it handles

maybe:

"conditions include being a fraction where the numerator is a sum of ...."

so you see you can include if it's a + node and what the first and second argument is etc into a much more concise language, and that's what I was hoping for.

"Check that the number of arguments in parent node is 2" is already implied in it being a fraction

e.g. several of your conditions can be combined into "the denominator has to be an addition/subtraction of a polynomial term and a constant term"

Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I love that you ahve this summary here though - it'll be so helpful!

File: lib/checks/canFindDenominatorInNumerator.js [hunk 23]
```diff
@@ -20,41 +33,70 @@ function canFindDenominatorInNumerator(node) {
   if (Node.Type.isParenthesis(denominator)) {
     denominator = denominator.content;
   }
-  if (!(numerator.op === '+' || numerator.op === '-' ||
-        denominator.op === '+' || numerator.op === '-')) {
+
+  let numeratorArgsLength;
+  // If numerator is '*' op, it signifies a single 'ax', should be assigned a
+  // length of 2
```
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: I think a clearer way to do this check is to see if it's a polynomial term (and then length 1) or else make it `numerator.args.length`

also if numerator doesn't have args, you should return false, which you don't do now right?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 19 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: Yeah, you're right. Oh wow, I didn't see isPolynomialTerm. Shoot, sorry for missing that. On a kind of unrelated note, to me it would make more sense if isPolynomialTerm was in the Type file instead, what do you think?

File: lib/checks/canFindDenominatorInNumerator.js (hunk 21 above)
Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: okay i also added another two lines of explanation to try to make it a bit more clear, let me know if you think it adds to the confusion or helps.

File: lib/checks/canFindDenominatorInNumerator.js (hunk 19 above)
Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: yeah that would be good to add there if it's easy! probably not instead, since it's nice to have the logic within polynomial term code, but if you could reference it in `Type` for now that'd be awesome

Comment (eltonlaw) [inline lib/checks/canFindDenominatorInNumerator.js]: uhh, actually maybe not. If we require 'PolynomialTerm' from 'Type', then they're referencing each other. So the only way to put it in there is if we just have two copies. So yeah maybe screw it.

Comment (evykassirer) [inline lib/checks/canFindDenominatorInNumerator.js]: ahh yeah okay I guess that's why that wasn't there then - ah well

File: test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js [hunk 24]
```diff
@@ -11,6 +11,13 @@ describe('breakUpNumerator', function() {
     ['(x+3+y)/3', '(x / 3 + 3/3 + y / 3)'],
     ['(2+x)/4', '(2/4 + x / 4)'],
     ['2(x+3)/3', '2 * (x / 3 + 3/3)'],
+    ['(2x + 3)/(2x + 2)', '((2x + 2) / (2x + 2) + 1 / (2x + 2))'],
+    ['(2x - 3)/(2x + 2)', '((2x + 2) / (2x + 2) - 5 / (2x + 2))'],
+    ['(2x + 3)/(2x)', '(2x / (2x) + 3 / (2x))'],
+    ['(3 + 2x)/(2x)', '(3 / (2x) + 2x / (2x))'],
+    ['(4x + 3)/(2x + 2)', '(2 * (2x + 2) / (2x + 2) - 1 / (2x + 2))'],
+//    ['(2x)/(3 + 2x)', '((3 + 2x) / (3 + 2x) - 3 / (3 + 2x))'],
+//    ['(2x)/(2x + 3)', '((2x + 3) / (2x + 3)) - 3 / (2x + 3)'],
```
Comment (evykassirer) [inline test/simplifyExpression/breakUpNumeratorSearch/breakUpNumeratorSearch.test.js]: if these tests pass, uncomment them

otherwise leave a TODO with what we'll need to do in the future to make them pass

File: test/checks/checks.test.js [hunk 25]
```diff
@@ -29,3 +29,26 @@ describe('canSimplifyPolynomialTerms addition', function() {
...
+  const tests = [
+    ['(x+1)/(x-2)', true],
+    ['(2x)/(x+4)', true],
+    ['(x)/(x+4)', true],
+    ['(x)/(2x+4)', true],
+    ['(x+3)/(x)', false], // Normal breakup function already solves this
+    ['(2x + 3)/(2x + 2)', true],
+    ['(2x+3)/(2x)', false], // Normal breakup function already solves this
+    ['(2x)/(2x + 2)', true],
+    ['(5x + 3)/(4)', false], // Normal breakup function already solves this
```
Comment (evykassirer) [inline test/checks/checks.test.js]: these comments are super helpful!

**Comment by eltonlaw** [comment]
Hey @aliang8 and @evykassirer, I think the problem with 2x / (2x +3) is the following line in breakUpNumeratorSearch/index.js:  
```
if (!Node.Type.isOperator(numerator) || numerator.op !== '+') {
    return Node.Status.noChange(node);
}
```
the second part numerator.op !== '+' evaluates to true so the no change node is returned. I was thinking about adding a part before this where it adds an op to  add a constant node '0' to the numerator in the case of a single polynomial numerator. But idk, it might end up really fragile. What do you think we should do?

**Comment by aliang8** [comment]
Ah good call, I see where you're referring to. Maybe you could add in a check to see if the numerator is a single polynomial term or if it is just a constant and handle it accordingly inside that if statement. By doing this, we can restrict to only the cases that we handle.

So in practice that if statement could be: `if ((!isOperator(node) || (numerator.op != '+')) && notSinglePolynomialTerm(node)) {return}` It seems a bit clunky as of now, but hopefully you could think of a better way to handle this.

**Comment by evykassirer** [comment]
What about if you check your own conditions first, then check 
`if (!Node.Type.isOperator(numerator) || numerator.op !== '+') {`
if your conditions don't pass?

**Comment by eltonlaw** [comment]
hey @evykassirer @aliang8, I've been trying to debug this but I can't seem to identify the problem. I've been poking around but I just can't seem to make sense of what's going on. When I try to incorporate the new checks everything seems to blow up. Sometimes tests from addConstantFractions give out errors, other times a bunch of tests in solveEquation fail. I am baffled. Do you guys want to check it out? 

In the meantime, I just commented it out and left a TODO.
